### ISSUE TITLE
[QST] difficulties with the Python example

### ISSUE THREAD

**Issue by pommedeterresautee** [NVIDIA/cutlass#575]
I am trying to make the Python example work on my 3090 RTX (capacity 86).
To make the Python example script work, it had to be modified to include `cuda` path in the `includes` list (not sure why it's not done by default).
I have 2 questions:
* how I am supposed to retrieve the result of the operation? Below I tried `cupy` and a simple device to host transfer, both gives the same tensor and it's not the expected output
* when I use `cutlass_simt_sgemm_256x128_8x5_nt_align1` (and the `GenerateSM80` generator), the script doesn't work anymore, it runs but tensor D is not modified (and `cupy` crashes)

As you may have guessed, I am discovering cutlass and may miss some obvious things.

```python
from cuda.cuda import CUdeviceptr, IntEnum

# System modules
import numpy as np

# CUDA Python modules
from cuda import cuda

# CUTLASS modules
import manifest as cutlass_manifest
import generator as generator
import rt as rt
import cupy as cp
import torch


#
# Construct an SGEMM
#


device_info = torch.cuda.get_device_properties(0)
manifest = cutlass_manifest.Manifest()
manifest.compute_capabilities = [device_info.major * 10 + device_info.minor]
generator.GenerateSM50_Simt(manifest, "11.7.0")
print(list(manifest.operations_by_name.keys()))
#
# Construct a GEMM operation
#
# for 50: 'cutlass_simt_sgemm_128x128_8x2_nt_align1'
# for 86: 'cutlass_simt_sgemm_256x128_8x5_nt_align1'
operation = manifest.operations_by_name[
    'cutlass_simt_sgemm_128x128_8x2_nt_align1']

#
# Construct a runtime GEMM operation
#
gemm = rt.Gemm(operation)

#
# Initialize context
#
err, = cuda.cuInit(0)  # should be called first

if err != cuda.CUresult.CUDA_SUCCESS:
    raise RuntimeError("CUDA Error %s" % str(err))

err, device = cuda.cuDeviceGet(0)

if err != cuda.CUresult.CUDA_SUCCESS:
    raise RuntimeError("CUDA Error %s" % str(err))

err, context = cuda.cuCtxCreate(0, device)

if err != cuda.CUresult.CUDA_SUCCESS:
    raise RuntimeError("CUDA Error %s" % str(err))

#
# Construct a module
#

include_paths = [
    '/home/geantvert/workspace/cutlass/include',
    '/home/geantvert/workspace/cutlass/tools/util/include',
    '/usr/local/cuda-11.7/targets/x86_64-linux/include/',
]

compilation_options = rt.CompilationOptions(architectures=[device_info.major * 10 + device_info.minor],
                                            include_paths=include_paths)

module = rt.Module('module.cu', [gemm], compilation_options)

#
# Setup a workspace
#

M, N, K = (2, 5, 3)

tensor_A = np.ones((M, K), dtype=np.float32)
tensor_B = np.ones((N, K), dtype=np.float32) * 2.0
tensor_C = np.ones((M, N), dtype=np.float32) + 1.0
tensor_D = np.zeros((M, N), dtype=np.float32)

err, tensor_A_d = cuda.cuMemAlloc(tensor_A.size * tensor_A.itemsize)  # type: IntEnum, CUdeviceptr
if err != cuda.CUresult.CUDA_SUCCESS:
    raise RuntimeError("CUDA Error %s" % str(err))

err, tensor_B_d = cuda.cuMemAlloc(tensor_B.size * tensor_B.itemsize)
if err != cuda.CUresult.CUDA_SUCCESS:
    raise RuntimeError("CUDA Error %s" % str(err))

err, tensor_C_d = cuda.cuMemAlloc(tensor_C.size * tensor_C.itemsize)
if err != cuda.CUresult.CUDA_SUCCESS:
    raise RuntimeError("CUDA Error %s" % str(err))

err, tensor_D_d = cuda.cuMemAlloc(tensor_D.size * tensor_D.itemsize)
if err != cuda.CUresult.CUDA_SUCCESS:
    raise RuntimeError("CUDA Error %s" % str(err))

err, stream = cuda.cuStreamCreate(0)
if err != cuda.CUresult.CUDA_SUCCESS:
    raise RuntimeError("CUDA Error %s" % str(err))

tensors = [
    (tensor_A_d, tensor_A),
    (tensor_B_d, tensor_B),
    (tensor_C_d, tensor_C),
    (tensor_D_d, tensor_D)
]

for tensor_device, tensor_host in tensors:
    data_size = tensor_host.size * tensor_host.itemsize
    print("Tensor has dimensions: %s (%d bytes)" % (str(tensor_host.size), tensor_host.itemsize))
    err, = cuda.cuMemcpyHtoD(tensor_device, tensor_host, data_size)
    print("updating tensor in device memory ", hex(int(tensor_device)))
    if err != cuda.CUresult.CUDA_SUCCESS:
        raise RuntimeError('CUDA Error %s' % str(err))

# Initialize a host buffer
#

arguments = rt.GemmArguments()

arguments.problem_size = rt.GemmCoord(M, N, K)

arguments.A = rt.TensorRef(tensor_A_d, M)
arguments.B = rt.TensorRef(tensor_B_d, N)
arguments.C = rt.TensorRef(tensor_C_d, M)
arguments.D = rt.TensorRef(tensor_D_d, M)

host_workspace = bytearray(gemm.get_host_workspace_size(arguments))
device_workspace = None

launch_config = gemm.plan(arguments)

byte_count = gemm.initialize(host_workspace, device_workspace, launch_config, arguments)

#
# Launch the kernel
#

err = gemm.run(host_workspace, device_workspace, launch_config)

if err != cuda.CUresult.CUDA_SUCCESS:
    raise RuntimeError('CUDA Error %s' % str(err))

#
# Verify results
#
err, = cuda.cuStreamSynchronize(stream)

if err != cuda.CUresult.CUDA_SUCCESS:
    raise RuntimeError("CUDA Error %s" % str(err))

#
# Debug reporting of byte array contents
#

def PrintBytearray(host_workspace):
  uint_str = None
  prefix = None
  print("uint32_t host_workspace[] = {")
  for idx, byte in enumerate(host_workspace):
    if not (idx % 4):
      if uint_str is not None:
        print(prefix, uint_str, ",")
      prefix = "/* offset: %d B */    0x" % idx
      uint_str = ""
    uint_str = "{:02x}".format(byte) + uint_str
  print("};")


PrintBytearray(host_workspace)
# err, tensor_D_d = cuda.cuMemAlloc(tensor_D.size * tensor_D.itemsize)

memory = cp.cuda.UnownedMemory(int(tensor_D_d), 0, owner=1)
memory_ptr = cp.cuda.MemoryPointer(memory, 0)
cp_array = cp.ndarray(shape=tensor_D.shape, memptr=memory_ptr, dtype=np.float32)
print(cp_array)


cuda.cuMemcpyDtoH(tensor_D.ctypes.data, int(tensor_D_d), tensor_D.size * tensor_D.itemsize)
print(tensor_D)


# to try
# https://github.com/tjyuyao/cutex

```

**Comment by apuaaChen** [comment]
The question 1 can be addressed directly with [cuda-python](https://nvidia.github.io/cuda-python/module/cuda.html?highlight=cumemcpydtoh#cuda.cuda.cuMemcpyDtoH):
```python
err, = cuda.cuMemcpyDtoH(tensor_D, tensor_D_d, tensor_D.size * tensor_D.itemsize)
if err != cuda.CUresult.CUDA_SUCCESS:
   raise RuntimeError("CUDA Error %s" % str(err))

err, = cudart.cudaDeviceSynchronize()

if err != cuda.CUresult.CUDA_SUCCESS:
  raise RuntimeError("CUDA Error %s" % str(err))
```

For question 2, the current script only works for a few cases. The reason is that in the file below, the params are hardcoded, whereas they should be configured based on the kernel configuration and the input logical arguments. There are other missing things such as setting maximum shared memory capacity. 

https://github.com/NVIDIA/cutlass/blob/master/tools/library/scripts/rt.py?plain=1#L489

Currently we are working on a python interface that addresses your concerns. Hopefully it can be merges in a few weeks.

**Comment by pommedeterresautee** [comment]
Thank you for your fast answer.

Regarding the retrieved tensor, it's the same as the one I retrieved with the code I posted below (both CUPY and cuMemcpyDtoH).

The output is:

```
[[6. 6. 6. 6. 6.]
 [6. 6. 0. 0. 0.]]
```

I was expecting the same shape but 6 everywhere (with the parameter I gave). 
Is it because of the hard-coded value not matching my hardware? (some layout thing?)

Regarding the ongoing work on a better Python interface, it would be terrific!

If I may, one thing a cutlass learner like me would enjoy is that the Python script example provides your "best of" examples in Python, like (IMHO) simple gemm, grouped gemm, some advanced epilogue stuff like softmax, relu or whatever in the same spirit. And everywhere it would go up to retrieve the output and compare with the expected one :-)

**Comment by apuaaChen** [comment]
> ved tensor, it's the same as the one I retrieved with the code I posted below (both CUPY and cuMemcpyDtoH).
> 
> The output is:

The wrong result is likely caused by the wrong parameters. The parameter of kernels in CUTLASS is generated in this way. First, user provides a logical argument like this one

https://github.com/NVIDIA/cutlass/blob/master/include/cutlass/gemm/kernel/gemm_universal.h?plain=1#L104-L241

Then, the argument is used to construct the params like

https://github.com/NVIDIA/cutlass/blob/master/include/cutlass/gemm/kernel/gemm_universal.h?plain=1#L308-L339

As you can see, the params is determined by both user-provided logical argument and the template arguments of the kernel.

The ongoing python interface would be like what you described. We are also exploring ways to enable user-defined epilogue functors

**Comment by pommedeterresautee** [comment]
Thank you it's clear to me. Will close the issue and wait for the new python examples :-)
